File: core/src/main/scala/diplomacy/variant/standard/rule/RetreatOrderGenerator.scala
package diplomacy.variant.standard.rule

import diplomacy.rule.InvalidOrderMessage

class RetreatOrderGenerator extends OrderGenerator {
  def ordersToSkipPhase(board: Board): Option[Set[Order]] = {
    val dislodgedUnits = board.unitStatuses collect {
      case (unit, x: UnitStatus.Dislodged[Power]) => unit -> x.attackedFrom // TODO
    }
    if (
      dislodgedUnits forall {
        case (unit, attackedFrom) =>
          RetreatPhaseUtils.locationsToRetreat(board)(unit, attackedFrom).isEmpty
      }
    ) {
      Option(
        (dislodgedUnits map { case (unit, _) => Order.Disband(unit) })(collection.breakOut)
      )
    } else {
      None
    }
  }
  def defaultOrder(board: Board)(unit: DiplomacyUnit): Order = Order.Disband(unit)
}

File: src/standardRule/retreat-order-generator.ts
import { OrderGenerator } from "./order-generator"
import { Utils } from "./utils"
import { Disband } from "./order"
import { Board, Unit } from "./types"

export class RetreatOrderGenerator<Power> implements OrderGenerator<Power> {
  ordersToSkipPhase (board: Board<Power>) {
    if (board.unitStatuses.size === 0) {
      return new Set()
    }
    if (
      [...board.unitStatuses].every(elem => {
        const [unit, status] = elem
        const locations = Utils.locationsToRetreat(board, unit, status.attackedFrom)
        return locations.size === 0
      })
    ) {
      // Disband all dislodged units
      const retval = new Set()
      board.unitStatuses.forEach((status, unit) => {
        retval.add(new Disband(unit))
      })
      return retval
    }
    return null
  }
  defaultOrderOf (board: Board<Power>, unit: Unit<Power>) {
    return new Disband(unit)
  }
}
